### DISCUSSION TITLE
Temporary Variables in forall scope

### DISCUSSION THREAD

**Discussion by SunandanAdhikary** [Z3Prover/z3#5088] [Q&A] [answered]
I am stuck with a forall statement. Found this answer useful. So just wanted to know, is there any way to define a temporary variable to use in the forall scope like we do in z3 using let keyword? Because if there are lots of variables with long expressions, it is difficult to parse for z3py it seems!

> Yeah, I don't think there's anything like `((y Int))` in the Python API (when creating your forall); this is because Python (not Z3!) needs to resolve the symbols you want to use inside of your quantification. This is 100% a parse-time thing and not a Z3 limitation.
> 
> Your example is something like:
> 
> ```python
> from z3 import *
> 
> s = Solver()
> 
> z = Int("z")
> y = Int("y")
> 
> s.add(ForAll([y], Implies(And(y > -5, y < 5), y + z > 10)))
> 
> print(s.sexpr())
> ```
> 

_Originally posted by @andrewvaughanj in https://github.com/Z3Prover/z3/issues/5044#issuecomment-784036442_

**Comment by aytey** [comment, marked as answer]
You don't have to write the constraint for `ForAll` "inline"; you can totally do something like:

```python
from z3 import *


def my_fun(y, z):
    return Implies(And(y > -5, y < 5), y + z > 10)


s = Solver()

z = Int("z")
y = Int("y")

s.add(ForAll([y], my_fun(y, z)))

print(s.sexpr())
```

I think in Z3 parlance, `my_fun` is called a "generator expression" -- that is, it a function that generates further Z3 expressions.

**Comment by SunandanAdhikary** [comment]
So I can write anything in place of `Implies... `in the function definition? Like we normally do in python? And reuse that?

**Comment by aytey** [comment]
Yes, as long as "normally do" returns a (self-contained) Z3 expression.

**Comment by SunandanAdhikary** [comment]
Understood. Thank you.
